junit5/src/main/java/org/jboss/weld/junit5/auto/ExcludedBeansExtension.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * JBoss, Home of Professional Open Source
+ * Copyright 2018, Red Hat, Inc., and individual contributors
+ * by the @authors tag. See the copyright.txt in the distribution for a
+ * full listing of individual contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.weld.junit5.auto;
+
+import javax.enterprise.context.NormalScope;
+import javax.enterprise.event.Observes;
+import javax.enterprise.inject.spi.Extension;
+import javax.enterprise.inject.spi.ProcessAnnotatedType;
+import javax.enterprise.inject.spi.WithAnnotations;
+import javax.inject.Scope;
+import java.lang.reflect.Type;
+import java.util.Set;
+
+
+/**
+ * Extension class that ensures selected classes are excluded as
+ * beans.
+ */
+public class ExcludedBeansExtension implements Extension {
+
+    private Set<Type> excludedBeanTypes;
+
+    ExcludedBeansExtension(Set<Type> excludedBeanTypes) {
+        this.excludedBeanTypes = excludedBeanTypes;
+    }
+
+    <T> void excludeBeans(@Observes @WithAnnotations({Scope.class, NormalScope.class}) ProcessAnnotatedType<T> pat) {
+        if (excludedBeanTypes.contains(pat.getAnnotatedType().getBaseType())) {
```

Comment:
If you alter this to do the check bases on `pat.getAnnotatedType().getTypeClosure()`, then the exclusions will operate (better) on interfaces. As in, you can then use `@ExcludeBeans` on an interface class and still get implementations vetoed. It will also help avoid some weird situations with resolution problems if user defines a producer that has interface as return type. Something like this is what I mean:

       Set<Type> typeClosure = new HashSet<>();
        typeClosure.addAll(pat.getAnnotatedType().getTypeClosure());
        typeClosure.retainAll(excludedBeanTypes);
        if (!typeClosure.isEmpty()) {
            pat.veto();
        }

Reply:
Agreed.